File: lib/product/sarif.ml
let level = function
  | Diagnostic.Critical | Error -> "error"
  | Warning -> "warning"
  | Note -> "note"

let region span =
  Json.Object
    [
      ("endColumn", Json.Int span.Span.stop.column);
      ("endLine", Json.Int span.stop.line);
      ("startColumn", Json.Int span.start.column);
      ("startLine", Json.Int span.start.line);
    ]

let physical_location span =
  Json.Object
    [
      ( "artifactLocation",
        Json.Object
          [ ("uri", Json.String (Util.normalize_slashes span.Span.file)) ] );
      ("region", region span);
    ]

let location span = Json.Object [ ("physicalLocation", physical_location span) ]

let rule_descriptor diagnostic =
  Json.Object
    [
      ("id", Json.String diagnostic.Diagnostic.rule_id);
      ("name", Json.String diagnostic.rule_id);
      ( "helpUri",
        Json.String
          ("https://workflow-verifier.dev/rules/"
          ^ String.lowercase_ascii diagnostic.rule_id) );
      ( "shortDescription",
        Json.Object [ ("text", Json.String diagnostic.message) ] );
    ]

let trace_location index (hop : Diagnostic.trace_hop) =
  Json.Object
    [
      ( "location",
        Json.Object
          [
            ("message", Json.Object [ ("text", Json.String hop.label) ]);
            ("physicalLocation", physical_location hop.span);
          ] );
      ("nestingLevel", Json.Int index);
    ]

let code_flows trace =
  Json.Array
    [
      Json.Object
        [
          ( "threadFlows",
            Json.Array
              [
                Json.Object
                  [ ("locations", Json.Array (List.mapi trace_location trace)) ];
              ] );
        ];
    ]

let fix_json (fix : Diagnostic.fix) =
  let artifact_changes =
    match (fix.replacement, fix.span) with
    | Some replacement, Some span ->
        Json.Array
          [
            Json.Object
              [
                ( "artifactLocation",
                  Json.Object
                    [
                      ( "uri",
                        Json.String (Util.normalize_slashes span.Span.file) );
                    ] );
                ( "replacements",
                  Json.Array
                    [
                      Json.Object
                        [
                          ("deletedRegion", region span);
                          ( "insertedContent",
                            Json.Object [ ("text", Json.String replacement) ] );
                        ];
                    ] );
              ];
          ]
    | _ -> Json.Array []
  in
  Json.Object
    [
      ("artifactChanges", artifact_changes);
      ("description", Json.Object [ ("text", Json.String fix.description) ]);
      ("properties", Json.Object [ ("kind", Json.String fix.kind) ]);
    ]

let result diagnostic =
  let fields =
    [
      ("codeFlows", code_flows diagnostic.Diagnostic.trace);
      ("level", Json.String (level diagnostic.Diagnostic.severity));
      ("locations", Json.Array [ location diagnostic.span ]);
      ("message", Json.Object [ ("text", Json.String diagnostic.message) ]);
      ( "partialFingerprints",
        Json.Object [ ("workflowVerifier/v1", Json.String diagnostic.id) ] );
      ( "properties",
        Json.Object
          [
            ( "confidence",
              Json.String (Diagnostic.confidence_name diagnostic.confidence) );
            ( "capabilities",
              Json.Array
                (List.map
                   (fun capability ->
                     Json.String (Ir.capability_name capability))
                   diagnostic.capabilities) );
            ( "evidence",
              Json.Array
                (List.map (fun value -> Json.String value) diagnostic.evidence)
            );
            ("diagnosticId", Json.String diagnostic.id);
          ] );
      ("ruleId", Json.String diagnostic.rule_id);
    ]
  in
  let fields =
    match diagnostic.fix with
    | None -> fields
    | Some fix -> ("fixes", Json.Array [ fix_json fix ]) :: fields
  in
  Json.Object fields

let to_json report =
  let diagnostics = Report.diagnostics report in
  let rules =
    diagnostics
    |> List.sort_uniq (fun left right ->
        String.compare left.Diagnostic.rule_id right.rule_id)
    |> List.map rule_descriptor
  in
  Json.Object
    [
      ( "$schema",
        Json.String
          "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json"
      );
      ( "runs",
        Json.Array
          [
            Json.Object
              [
                ( "automationDetails",
                  Json.Object [ ("id", Json.String report.Report.digest) ] );
                ( "invocations",
                  Json.Array
                    [
                      Json.Object
                        [
                          ( "executionSuccessful",
                            Json.Bool
                              (report.provenance.exit_code <> 4
                              && report.provenance.exit_code <> 5) );
                          ("exitCode", Json.Int report.provenance.exit_code);
                          ( "exitCodeDescription",
                            Json.String
                              (Report.gate_result_name
                                 report.provenance.gate_result) );
                          ( "properties",
                            Json.Object
                              [
                                ("reportDigest", Json.String report.digest);
                                ( "sourceManifestDigest",
                                  Json.String
                                    report.provenance.source_manifest_digest );
                              ] );
                        ];
                    ] );
                ("results", Json.Array (List.map result diagnostics));
                ( "tool",
                  Json.Object
                    [
                      ( "driver",
                        Json.Object
                          [
                            ( "informationUri",
                              Json.String
                                "https://github.com/P4suta/workflow-verifier" );
                            ("name", Json.String "workflow-verifier");
                            ("rules", Json.Array rules);
                            ("version", Json.String report.Report.tool_version);
                          ] );
                    ] );
              ];
          ] );
      ("version", Json.String "2.1.0");
    ]

let to_canonical_json report = Json.to_string (to_json report) ^ "\n"

File: crates/product/src/sarif.rs
use crate::{Report, TOOL_NAME};
use std::collections::{BTreeMap, BTreeSet};
use workflow_verifier_foundation::{JsonValue, Span, normalize_slashes};
use workflow_verifier_verifier::{Diagnostic, Fix, Severity, TraceHop};

#[must_use]
pub fn report_to_sarif(report: &Report) -> String {
    let diagnostics = report.diagnostics();
    let mut seen = BTreeSet::new();
    let rules = diagnostics
        .iter()
        .filter(|diagnostic| seen.insert(diagnostic.rule_id.clone()))
        .map(rule_descriptor)
        .collect();
    let run = JsonValue::Object(BTreeMap::from([
        (
            "automationDetails".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "id".to_owned(),
                JsonValue::String(report.digest.clone()),
            )])),
        ),
        (
            "invocations".to_owned(),
            JsonValue::Array(vec![invocation(report)]),
        ),
        (
            "results".to_owned(),
            JsonValue::Array(diagnostics.iter().map(result).collect()),
        ),
        ("tool".to_owned(), tool(report, rules)),
    ]));
    JsonValue::Object(BTreeMap::from([
        (
            "$schema".to_owned(),
            JsonValue::String(
                "https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json"
                    .to_owned(),
            ),
        ),
        ("runs".to_owned(), JsonValue::Array(vec![run])),
        ("version".to_owned(), JsonValue::String("2.1.0".to_owned())),
    ]))
    .canonical_line()
}

fn invocation(report: &Report) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "executionSuccessful".to_owned(),
            JsonValue::Boolean(!matches!(report.provenance.exit_code, 4 | 5)),
        ),
        (
            "exitCode".to_owned(),
            JsonValue::Integer(report.provenance.exit_code),
        ),
        (
            "exitCodeDescription".to_owned(),
            JsonValue::String(report.provenance.gate_result.name().to_owned()),
        ),
        (
            "properties".to_owned(),
            JsonValue::Object(BTreeMap::from([
                (
                    "reportDigest".to_owned(),
                    JsonValue::String(report.digest.clone()),
                ),
                (
                    "semanticDigest".to_owned(),
                    JsonValue::String(report.semantic_digest.clone()),
                ),
                (
                    "sourceManifestDigest".to_owned(),
                    JsonValue::String(report.provenance.source_manifest_digest.clone()),
                ),
            ])),
        ),
    ]))
}

fn tool(report: &Report, rules: Vec<JsonValue>) -> JsonValue {
    JsonValue::Object(BTreeMap::from([(
        "driver".to_owned(),
        JsonValue::Object(BTreeMap::from([
            (
                "informationUri".to_owned(),
                JsonValue::String("https://github.com/P4suta/workflow-verifier".to_owned()),
            ),
            ("name".to_owned(), JsonValue::String(TOOL_NAME.to_owned())),
            ("rules".to_owned(), JsonValue::Array(rules)),
            (
                "version".to_owned(),
                JsonValue::String(report.tool_version.clone()),
            ),
        ])),
    )]))
}

fn rule_descriptor(diagnostic: &Diagnostic) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "helpUri".to_owned(),
            JsonValue::String(format!(
                "https://workflow-verifier.dev/rules/{}",
                diagnostic.rule_id.to_ascii_lowercase()
            )),
        ),
        (
            "id".to_owned(),
            JsonValue::String(diagnostic.rule_id.clone()),
        ),
        (
            "name".to_owned(),
            JsonValue::String(diagnostic.rule_id.clone()),
        ),
        (
            "shortDescription".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "text".to_owned(),
                JsonValue::String(diagnostic.message.clone()),
            )])),
        ),
    ]))
}

fn result(diagnostic: &Diagnostic) -> JsonValue {
    let mut fields = BTreeMap::from([
        (
            "codeFlows".to_owned(),
            JsonValue::Array(vec![JsonValue::Object(BTreeMap::from([(
                "threadFlows".to_owned(),
                JsonValue::Array(vec![JsonValue::Object(BTreeMap::from([(
                    "locations".to_owned(),
                    JsonValue::Array(
                        diagnostic
                            .trace
                            .iter()
                            .enumerate()
                            .map(|(index, hop)| trace_location(index, hop))
                            .collect(),
                    ),
                )]))]),
            )]))]),
        ),
        (
            "level".to_owned(),
            JsonValue::String(level(diagnostic.severity).to_owned()),
        ),
        (
            "locations".to_owned(),
            JsonValue::Array(vec![location(&diagnostic.span)]),
        ),
        (
            "message".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "text".to_owned(),
                JsonValue::String(diagnostic.message.clone()),
            )])),
        ),
        (
            "partialFingerprints".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "workflowVerifier/v1".to_owned(),
                JsonValue::String(diagnostic.id.clone()),
            )])),
        ),
        ("properties".to_owned(), result_properties(diagnostic)),
        (
            "ruleId".to_owned(),
            JsonValue::String(diagnostic.rule_id.clone()),
        ),
    ]);
    if let Some(fix) = &diagnostic.fix {
        fields.insert("fixes".to_owned(), JsonValue::Array(vec![fix_json(fix)]));
    }
    JsonValue::Object(fields)
}

fn result_properties(diagnostic: &Diagnostic) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "capabilities".to_owned(),
            JsonValue::Array(
                diagnostic
                    .capabilities
                    .iter()
                    .map(|value| JsonValue::String(value.name().to_owned()))
                    .collect(),
            ),
        ),
        (
            "confidence".to_owned(),
            JsonValue::String(diagnostic.confidence.name().to_owned()),
        ),
        (
            "diagnosticId".to_owned(),
            JsonValue::String(diagnostic.id.clone()),
        ),
        (
            "evidence".to_owned(),
            JsonValue::Array(
                diagnostic
                    .evidence
                    .iter()
                    .cloned()
                    .map(JsonValue::String)
                    .collect(),
            ),
        ),
    ]))
}

fn level(severity: Severity) -> &'static str {
    match severity {
        Severity::Critical | Severity::Error => "error",
        Severity::Warning => "warning",
        Severity::Note => "note",
    }
}

fn region(span: &Span) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "endColumn".to_owned(),
            JsonValue::Integer(i64::from(span.stop.column)),
        ),
        (
            "endLine".to_owned(),
            JsonValue::Integer(i64::from(span.stop.line)),
        ),
        (
            "startColumn".to_owned(),
            JsonValue::Integer(i64::from(span.start.column)),
        ),
        (
            "startLine".to_owned(),
            JsonValue::Integer(i64::from(span.start.line)),
        ),
    ]))
}

fn physical_location(span: &Span) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "artifactLocation".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "uri".to_owned(),
                JsonValue::String(normalize_slashes(&span.file)),
            )])),
        ),
        ("region".to_owned(), region(span)),
    ]))
}

fn location(span: &Span) -> JsonValue {
    JsonValue::Object(BTreeMap::from([(
        "physicalLocation".to_owned(),
        physical_location(span),
    )]))
}

fn trace_location(index: usize, hop: &TraceHop) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        (
            "location".to_owned(),
            JsonValue::Object(BTreeMap::from([
                (
                    "message".to_owned(),
                    JsonValue::Object(BTreeMap::from([(
                        "text".to_owned(),
                        JsonValue::String(hop.label.clone()),
                    )])),
                ),
                ("physicalLocation".to_owned(), physical_location(&hop.span)),
            ])),
        ),
        (
            "nestingLevel".to_owned(),
            JsonValue::Integer(i64::try_from(index).unwrap_or(i64::MAX)),
        ),
    ]))
}

fn fix_json(fix: &Fix) -> JsonValue {
    let changes = match (&fix.replacement, &fix.span) {
        (Some(replacement), Some(span)) => vec![JsonValue::Object(BTreeMap::from([
            (
                "artifactLocation".to_owned(),
                JsonValue::Object(BTreeMap::from([(
                    "uri".to_owned(),
                    JsonValue::String(normalize_slashes(&span.file)),
                )])),
            ),
            (
                "replacements".to_owned(),
                JsonValue::Array(vec![JsonValue::Object(BTreeMap::from([
                    ("deletedRegion".to_owned(), region(span)),
                    (
                        "insertedContent".to_owned(),
                        JsonValue::Object(BTreeMap::from([(
                            "text".to_owned(),
                            JsonValue::String(replacement.clone()),
                        )])),
                    ),
                ]))]),
            ),
        ]))],
        _ => Vec::new(),
    };
    JsonValue::Object(BTreeMap::from([
        ("artifactChanges".to_owned(), JsonValue::Array(changes)),
        (
            "description".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "text".to_owned(),
                JsonValue::String(fix.description.clone()),
            )])),
        ),
        (
            "properties".to_owned(),
            JsonValue::Object(BTreeMap::from([(
                "kind".to_owned(),
                JsonValue::String(fix.kind.clone()),
            )])),
        ),
    ]))
}
